aemcmc/negative_binomial_horseshoe_gibbs.py

```diff
@@ -0,0 +1,199 @@
+        parameter of the horseshoe prior.
+    global_shrinkage_init: TensorVariable
+        A tensor describing the starting values of the global shrinkage
+        parameter of the horseshoe prior.
+    n_samples: int
+        The number of posterior samples to generate. Defaults to ``1000``.
+
+    Returns
+    -------
+    (outputs, updates): tuple
+        A  symbolic sescription of the sampling result to be used when
+        compiling a function to be used for sampling.
+
+    References
+    ----------
+    ..[1] Makalic, Enes & Schmidt, Daniel. (2015). A Simple Sampler for the
+          Horseshoe Estimator. 10.1109/LSP.2015.2503725.
+    """
+    beta = at.as_tensor_variable(beta_init, "beta", dtype=aesara.config.floatX)
+    if beta.ndim != 1:
+        raise ValueError(
+            "`beta_init` must be a vector whose length is equal to the "
+            "number of columns in the covariate matrix"
+        )
+
+    lambda2 = at.as_tensor_variable(
+        local_shrinkage_init, name="lambda2", dtype=aesara.config.floatX
+    )
+    if lambda2.ndim != 1:
+        raise ValueError(
+            "The local shrinkage initial value must be a vector whose size "
+            "is equal to the number of columns in the covariate matrix"
+        )
+
+    tau2 = at.as_tensor_variable(
+        global_shrinkage_init, name="tau2", dtype=aesara.config.floatX
+    )
+    if tau2.ndim != 0:
+        raise ValueError("The global shrinkage initial value must be a scalar")
+
+    sigma2 = at.empty_like(tau2)
+    sigma2.name = "sigma2"
+
+    upsilon = at.empty_like(beta)
+    upsilon.name = "upsilon"
+
+    zeta = at.empty_like(tau2)
+    zeta.name = "zeta"
+
+    def step_fn(
+        beta: TensorVariable,
+        sigma2: TensorVariable,
+        upsilon: TensorVariable,
+        zeta: TensorVariable,
+        lambda2: TensorVariable,
+        tau2: TensorVariable,
+        X: TensorVariable,
+        y: TensorVariable,
+        h: TensorVariable,
+    ):
+        """
+        Complete one full update of the gibbs sampler and return the new state
+        of the posterior conditional parameters.
+        """
+        xb = X @ beta
+        w = srng.gen(polyagamma, y + h, xb)
+
+        lambda2tau2 = lambda2 * tau2
+        sigma2_new = srng.invgamma(
+            0.5 * (X.shape[0] + X.shape[1]),
+            0.5 * ((y - xb).T @ (y - xb) + beta @ (beta / lambda2tau2)),
+        )
+
+        beta_new = _update_beta(srng, w, lambda2tau2, X, 0.5 * (y - h))
+
+        upsilon_new = srng.invgamma(1, 1 + 1 / lambda2)
+        zeta_new = srng.invgamma(1, 1 + 1 / tau2)
+
+        beta2 = beta_new * beta_new
+        lambda2_new = srng.invgamma(
+            1,
+            1 / upsilon_new + beta2 / (2 * tau2 * sigma2_new),
+        )
+        tau2_new = srng.invgamma(
+            0.5 * (X.shape[1] + 1),
+            1 / zeta_new + 0.5 * (beta2 / lambda2_new).sum() / sigma2_new,
+        )
```

Comment:
These steps can go in a general sampler step function (e.g. `normal_horseshoe_step`), but we still need to figure out what to do about the Polya-gamma part and how the two come together.

Ultimately, the "entry point" will be a model graph (e.g. an Aesara graph that contains `RandomVariable`s and represents the statistical model), and we will "parse" that and identify the observation model (that's fairly simple) and the regression relationship it has with a Horseshoe prior.  

In other words, we would get a graph constructed like `Y_rv = at.random.normal(X @ beta_rv, sigma, name="Y")` and `beta_rv = at.random.normal(0, at.sqrt(tau2_rv * lambda2_rv), name="beta")`, where `tau2_rv = at.random.halfcauchy(...)`, etc., and determine that `beta_rv` is a Horseshoe (e.g. via unification), then we would construct a sampler with the steps here.

During this same process, when `Y_rv` is instead an `at.random.negative_binomial` (parameterized in such a way that its mean is `X @ beta_rv`, of course), we would then employ the Polya-gamma currently in place.

We don't have to deal with the unification (i.e. "pattern matching") for all this right now, but we still need to make sure that everything is modular enough to use in such a framework.